tests/request_parsing_tests.rs

```diff
@@ -0,0 +1,187 @@
+fn test_create_window2() {
+    use x11rb::protocol::xproto::{CreateWindowAux, CreateWindowRequest, Gravity, WindowClass};
+    let header = RequestHeader {
+        major_opcode: 1,
+        minor_opcode: 0x18,
+        remaining_length: 11,
+    };
+    let body = &[
+        0x3, 0x0, 0x40, 0x5, 0x13, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1, 0x0, 0x0, 0x0,
+        0x1, 0x0, 0x5a, 0x3, 0x0, 0x0, 0x1a, 0x20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+        0x0, 0x1, 0x0, 0x0, 0x0, 0x2, 0x0, 0x40, 0x5,
+    ];
+    let r = CreateWindowRequest::try_parse_request(header, body).unwrap();
+    assert_eq!(
+        r,
+        CreateWindowRequest {
+            depth: 0x18,
+            wid: 0x05400003,
+            parent: 0x00000513,
+            x: 0,
+            y: 0,
+            width: 1,
+            height: 1,
+            border_width: 0,
+            class: WindowClass::InputOutput,
+            visual: 0x35a,
+            value_list: Cow::Owned(CreateWindowAux {
+                background_pixel: Some(0),
+                border_pixel: Some(0),
+                bit_gravity: Some(Gravity::NorthWest),
+                colormap: Some(0x05400002),
+                ..Default::default()
+            }),
+        }
+    );
+}
+
+#[test]
+fn test_change_window_attributes() {
+    use x11rb::protocol::xproto::{
+        ChangeWindowAttributesAux, ChangeWindowAttributesRequest, EventMask,
+    };
+    let header = RequestHeader {
+        major_opcode: 2,
+        minor_opcode: 0,
+        remaining_length: 3,
+    };
+    let body = &[0x13, 0x5, 0x0, 0x0, 0x0, 0x8, 0x0, 0x0, 0x0, 0x0, 0x40, 0x0];
+    let r = ChangeWindowAttributesRequest::try_parse_request(header, body).unwrap();
+    assert_eq!(
+        r,
+        ChangeWindowAttributesRequest {
+            window: 0x0513,
+            value_list: Cow::Owned(ChangeWindowAttributesAux {
+                event_mask: Some(EventMask::PropertyChange.into()),
+                ..Default::default()
+            }),
+        }
+    );
+}
+
+#[test]
+fn test_get_window_attributes() {
+    use x11rb::protocol::xproto::GetWindowAttributesRequest;
+    let header = RequestHeader {
+        major_opcode: 3,
+        minor_opcode: 0,
+        remaining_length: 1,
+    };
+    let body = &[0x2, 0x0, 0xe0, 0x0];
+    let r = GetWindowAttributesRequest::try_parse_request(header, body).unwrap();
+    assert_eq!(r, GetWindowAttributesRequest { window: 0x00e00002 });
+}
+
+#[test]
+fn test_get_input_focus() {
+    use x11rb::protocol::xproto::GetInputFocusRequest;
+    let header = RequestHeader {
+        major_opcode: 43,
+        minor_opcode: 0,
+        remaining_length: 0,
+    };
+    let body = &[];
+    let r = GetInputFocusRequest::try_parse_request(header, body).unwrap();
+    assert_eq!(r, GetInputFocusRequest,);
+}
+
+#[cfg(feature = "randr")]
+#[test]
+fn test_randr_get_output_property() {
+    use x11rb::protocol::randr::GetOutputPropertyRequest;
+    let header = RequestHeader {
```

Comment:
(Comment is at a random position) Why do we need `RequestHeader`?

I think that it might be possible to do without. The parsing code could just parse the request "normally" without skipping any fields. Opcodes can then be compared against expected values. Things that are not needed can be sent to `let _ =`.

`RequestHeader` does not bring anything to the parsing code, does it?

Reply:
The main reason I did it this way is that parsing requests is not stateless. You need to know the state of the connection with the server (in particular whether or not `BigRequests` is enabled and what extensions are enabled at what major opcodes) in order to correctly parse requests. We could choose to decide that we don't care about tracking whether `BigRequests` is enabled and simply always accept that format. However, even if we did that, someone outside of a `FooRequest::try_parse` implementation is always going to have to look at the first two bytes of the request to determine which parsing implementation to feed the buffer to (and for any opcodes above 127 they need to have been parsing the connection all along to know which extension opcodes exist). So `RequestHeader` is currently all the stuff that requires external state to process.

I'm not necessarily opposed to changing it but that is the motivation for the current design.

Reply:
This is kind of similar to `x11rb::protocol::Event` and `Error`. That code also needs to look up extensions based on error code / response type / major opcode. I guess we should also do something similar for request parsing (but that can be done in another PR).

This code uses some ad-hoc functions at the end of the file to extract the necessary info from the `&[u8]`.

A possible API for `x11rb::protocol::Request` would then be `fn parse(request: &[u8], ext_info_provider: &dyn ExtInfoProvider) -> Result<Self, ParseError>` (this is just the existing API for `Event`/`Error`).
Your version would be `-> Result<(RequestHeader, Self), ParseError>`, right? However... does the request header provide anything useful here? The request is already identified by the variant of the enum `Self` that is being returned. The request length is known to the caller as `.len()` of the provided slice. The difference between major/minor opcode does not really matter, I think.

What kind of API do you have in mind for this? Or did you plan to hand-write the necessary bits outside of x11rb?

Reply:
> in particular whether or not BigRequests is enabled 

Well... a big request is an illegal request when BigRequests is disabled, since a length of 0 would mean that there is not even enough space for the length field... So, personally I don't mind to assume that `BigRequests` is always enabled. Technically, this could also check the maximum request length, but this seems to be the wrong code to do that (since the request was already received at this point and is in-memory, such handling should be done before everything is read).

I also now looked at the X11 server and actually... there are some requests where it requires an exact length and sometimes where "trailing garbage" is okay, There are several macros for this in `include/dix.h`:
- `REQUEST_SIZE_MATCH`, 860
- `REQUEST_AT_LEAST_SIZE`, 346
- `REQUEST_AT_LEAST_EXTRA_SIZE`, 3
- `REQUEST_FIXED_SIZE`, 107

For some reason, `ProcNoOperation` uses `REQUEST_AT_LEAST_SIZE` and does not do an exact match. The number after the macro in the list above is the result of `git grep MACRO | wc -l`. Seems like: It's complicated

Reply:
Shorter way to parse length fields:
```rust
let (length, remaining) = match u16::try_parse(remaining)? {
  // BigRequests?
  (0, remaining) => u32::try_parse(remaining),
  (n, remaining) => (n.into(), remaining),
}
```

Reply:
Something else I realized is that regardless of our decision around `BigRequests` whoever is driving parsing has to parse the length themselves anyways to know if they have received enough bytes before calling the parsing method. Since `ParseError` doesn't distinguish between various types of errors "insufficient bytes" is otherwise indistinguishable from other types of parsing errors.

Reply:
Yup. That's also how the existing code in e.g. `RustConnection` parses errors/events/replies.

Once upon a time, the only possible parse error was "not enough data". Since then, a lot has happened. I wouldn't mind improving `ParseError`, but so far I do not really see a good use case for that. Our parsing errors most likely would cause memory bugs in anything with XCB, so apparently all of this "never happens" anyway...